### PR TITLE
Add parameters for alternate extensions and encrypt-format

### PR DESCRIPTION
**Pull request by vermashi**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs [hunk 1]
```diff
@@ -86,6 +86,22 @@ public class DisableAzureDiskEncryptionCommand : VirtualMachineExtensionBaseCmdl
             HelpMessage = "Disable auto-upgrade of minor version")]
         public SwitchParameter DisableAutoUpgradeMinorVersion { get; set; }
 
+        [Parameter(
+            Mandatory = false,
+            Position = 6,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The extension type. Specify this parameter to override its default value of \"AzureDiskEncryption\" for Windows VMs and \"AzureDiskEncryptionForLinux\" for Linux VMs.")]
+        [ValidateNotNullOrEmpty]
+        public string ExtensionType { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs]: @vermashi is there a closed set of values that can be provided to this parameter? If so, please use a `ValidateSet` to define the possible set of values to allow a user to tab through them

File: src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs [hunk 2]
```diff
@@ -86,6 +86,22 @@ public class DisableAzureDiskEncryptionCommand : VirtualMachineExtensionBaseCmdl
...
+
+        [Parameter(
+            Mandatory = false,
+            Position = 7,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The extension publisher name. Specify this parameter only to override the default value of \"Microsoft.Azure.Security\".")]
+        [ValidateNotNullOrEmpty]
+        public string ExtensionPublisherName { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs]: @vermashi same comment about possibly using a `ValidateSet`

File: src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/SetAzureDiskEncryptionExtension.cs [hunk 3]
```diff
@@ -174,6 +174,29 @@ public class SetAzureDiskEncryptionExtensionCommand : VirtualMachineExtensionBas
...
+            HelpMessage = "The extension publisher name. Specify this parameter only to override the default value of \"Microsoft.Azure.Security\".")]
+        [ValidateNotNullOrEmpty]
+        public string ExtensionPublisherName { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            Position = 18,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Encrypt-Format all data drives that are not already encrypted")]
+        public SwitchParameter EncryptFormatAll { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/SetAzureDiskEncryptionExtension.cs]: @vermashi a `SwitchParameter` should not have a position nor should it accept a value from the pipeline

File: src/ResourceManager/Compute/Commands.Compute/help/Disable-AzureRmVMDiskEncryption.md [hunk 4]
```diff
@@ -76,6 +78,32 @@ Accept pipeline input: True (ByPropertyName)
 Accept wildcard characters: False
 ```
 
+### -ExtensionPublisherName
+The extension publisher name. Specify this parameter only to override the default value of "Microsoft.Azure.Security".```yaml
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/help/Disable-AzureRmVMDiskEncryption.md]: @vermashi this is a bug in `platyPS` - please make sure the ` ```yaml ` string is on  a line by itself. See the other parameters in this file for guidance.

File: src/ResourceManager/Compute/Commands.Compute/help/Disable-AzureRmVMDiskEncryption.md [hunk 5]
```diff
@@ -76,6 +78,32 @@ Accept pipeline input: True (ByPropertyName)
...
+
+Required: False
+Position: Named
+Default value: None
+Accept pipeline input: True (ByPropertyName)
+Accept wildcard characters: False
+```
+
+### -ExtensionType
+The extension type. Specify this parameter to override its default value of "AzureDiskEncryption" for Windows VMs and "AzureDiskEncryptionForLinux" for Linux VMs.```yaml
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/help/Disable-AzureRmVMDiskEncryption.md]: @vermashi same comment

File: src/ResourceManager/Compute/Commands.Compute/help/Get-AzureRmVMDiskEncryptionStatus.md [hunk 6]
```diff
@@ -48,6 +50,32 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -ExtensionPublisherName
+The extension publisher name. Specify this parameter only to override the default value of "Microsoft.Azure.Security".```yaml
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/help/Get-AzureRmVMDiskEncryptionStatus.md]: @vermashi same comment in this file for the added parameters

File: src/ResourceManager/Compute/Commands.Compute/help/Set-AzureRmVMDiskEncryptionExtension.md [hunk 7]
```diff
@@ -293,6 +296,45 @@ Accept pipeline input: True (ByPropertyName)
 Accept wildcard characters: False
 ```
 
+### -EncryptFormatAll
+Encrypt-Format all data drives that are not already encrypted```yaml
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/help/Set-AzureRmVMDiskEncryptionExtension.md]: @vermashi same comment in this file for the added parameters

File: src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/AzureDiskEncryptionExtensionContext.cs [hunk 8]
```diff
@@ -25,10 +25,12 @@ public class AzureDiskEncryptionExtensionContext : PSVirtualMachineExtension
     {
         public const string LinuxExtensionDefaultPublisher = "Microsoft.Azure.Security";
         public const string LinuxExtensionDefaultName = "AzureDiskEncryptionForLinux";
+        public const string LinuxExtensionDefaultType = "AzureDiskEncryptionForLinux";
         public const string LinuxExtensionDefaultVersion = "0.1";
 
         public const string ExtensionDefaultPublisher = "Microsoft.Azure.Security";
         public const string ExtensionDefaultName = "AzureDiskEncryption";
+        public const string ExtensionDefaultType = "AzureDiskEncryption";
```
Comment (ejarvi) [inline src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/AzureDiskEncryptionExtensionContext.cs]: 👍  ... nice to see this getting disambiguated :)

File: src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs (hunk 2 above)
Comment (ejarvi) [reply src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs]: For background, the defaults won't change, and the user can get those if they don't specify this optional parameter.

File: src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs (hunk 1 above)
Comment (vermashi) [reply src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs]: No, users are not limited to a closed set of values for this parameter.

**Comment by vermashi** [comment]
Have addressed most of the comments. Working on the tests.

**Comment by vermashi** [comment]
@cormacpayne 
1) I have been trying to fix the tests but I'm running into an issue that when I try to create KeyVaults while running in Record Mode, I get the exception "TestManagementClientHelper class wasn't initialized with the ResourceManagementClient client". I'm not confident I can fix this issue before the snap date.
2) Our team has been considering that since the primary purpose of "ExtensionPublisherName" parameter will be to redirect the cmdlet to a private build of the extension, if its ok to pass this information, when required, via an EnvironmentVariable. 

TL;DR: 1- The tests might not get fixed by snap date, possible to ship without? Is it ok if we replace some of the newly added parameters in this PR with env-variables?

**Comment by maddieclayton** [comment]
@vermashi Please fix the merge conflicts and push.

File: src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs [hunk 9]
```diff
@@ -89,6 +89,22 @@ public class DisableAzureDiskEncryptionCommand : VirtualMachineExtensionBaseCmdl
             HelpMessage = "Disable auto-upgrade of minor version")]
         public SwitchParameter DisableAutoUpgradeMinorVersion { get; set; }
 
+        [Parameter(
+            Mandatory = false,
+            Position = 6,
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/DisableAzureDiskEncryption.cs]: @vermashi nit: the number of positional parameters in a single set should be capped at four or five. After this, it can be difficult for the user to remember the order in which they should provide the values, so our recommendation would be to remove the positions from the parameters you added (I would also include the `DisableAutoUpgradeMinorVersion` parameter above since switch parameters are only named ☺️ )

File: src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/GetAzureDiskEncryptionStatus.cs [hunk 10]
```diff
@@ -58,6 +58,22 @@ public class GetAzureDiskEncryptionStatusCommand : VirtualMachineExtensionBaseCm
         [ValidateNotNullOrEmpty]
         public string Name { get; set; }
         
+        [Parameter(
+            Mandatory = false,
+            Position = 3,
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/GetAzureDiskEncryptionStatus.cs]: @vermashi nit: you can decide whether or not the positions are needed for the new parameters ☺️

File: src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/SetAzureDiskEncryptionExtension.cs [hunk 11]
```diff
@@ -367,9 +392,17 @@ private Hashtable GetExtensionPublicSettings()
             publicSettings.Add(AzureDiskEncryptionExtensionConstants.keyVaultUrlKey, DiskEncryptionKeyVaultUrl ?? String.Empty);
             publicSettings.Add(AzureDiskEncryptionExtensionConstants.keyEncryptionKeyUrlKey, KeyEncryptionKeyUrl ?? String.Empty);
             publicSettings.Add(AzureDiskEncryptionExtensionConstants.volumeTypeKey, VolumeType ?? String.Empty);
-            publicSettings.Add(AzureDiskEncryptionExtensionConstants.encryptionOperationKey, AzureDiskEncryptionExtensionConstants.enableEncryptionOperation);
             publicSettings.Add(AzureDiskEncryptionExtensionConstants.sequenceVersionKey, SequenceVersion ?? String.Empty);
 
+            if (EncryptFormatAll)
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/SetAzureDiskEncryptionExtension.cs]: @vermashi nit: `EncryptFormatAll.IsPresent` for consistency with other cmdlets

File: src/ResourceManager/Compute/Commands.Compute/help/Set-AzureRmVMDiskEncryptionExtension.md [hunk 12]
```diff
@@ -294,6 +296,51 @@ Accept pipeline input: True (ByPropertyName)
...
+
+```yaml
+Type: SwitchParameter
+Parameter Sets: (All)
+Aliases: 
+
+Required: False
+Position: Named
+Default value: None
+Accept pipeline input: True (ByPropertyName)
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/help/Set-AzureRmVMDiskEncryptionExtension.md]: @vermashi it looks like the help for these cmdlets needs to be regenerated

**Comment by cormacpayne** [comment]
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/116/

**Comment by cormacpayne** [comment]
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/141/
